### ISSUE TITLE
`(...) -> NoReturn` doesn't work where `(...) -> None` is expected

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1048]
This example from `signal`'s docs (https://docs.python.org/3/library/signal.html#example) doesn't typecheck:
```py
import signal, os

def handler(signum, frame):
    print('Signal handler called with signal', signum)
    raise OSError("Couldn't open device!")

signal.signal(signal.SIGALRM, handler)
```
`handler` is expected to return `None`, but `handler` has the return type of `NoReturn` , so it doesn't fit there.

The error goes away when you make the return type annotation be `None`.
```py
def handler(signum, frame) -> None:
    print('Signal handler called with signal', signum)
    raise OSError("Couldn't open device!")
```

If `(...) -> NoReturn` is a subtype of `(...) -> None`, the example should have worked.
If `(...) -> NoReturn` isn't a subtype of `(...) -> None`, why is it possible to assign `-> None` to the function?

**Comment by erictraut** [comment]
The behavior is as designed. If you do not provide a return type declaration for a function, Pyright attempts to infer the return type. In this case, the inferred return type is `NoReturn` because there is no code path by which a value is returned. If you provide an explicit return type annotation, Pyright doesn't need to infer the return type. It simply verifies that all return statements honor the return type of the declared type. Since there are no return statements, it checks out.
